src/core/net/srp_server.cpp

```diff
@@ -0,0 +1,1515 @@
+
+            if (service->IsDeleted())
+            {
+                existingHost->RemoveService(existingService);
+            }
+            else
+            {
+                existingService = existingHost->AddService(service->GetFullName());
+
+                VerifyOrExit(existingService != nullptr, aError = OT_ERROR_NO_BUFS);
+                SuccessOrExit(aError = existingService->CopyResourcesFrom(*service));
+
+                otLogInfoSrp("SRP server: adds/updates service %s", existingHost->GetFullName());
+            }
+        }
+
+        Host::Destroy(&aHost);
+    }
+    else
+    {
+        otLogInfoSrp("SRP server: adds new host %s", aHost.GetFullName());
+
+        // Add current host in the SRP update if this is the first update of it.
+        AddHost(&aHost);
+    }
+
+    // Restart or stop the lease timer.
+    HandleLeaseTimer();
+
+exit:
+    if (aError == OT_ERROR_NONE && !(grantedLease == hostLease && grantedKeyLease == hostKeyLease))
+    {
+        SendResponse(aDnsHeader, grantedLease, grantedKeyLease, aMessageInfo);
+    }
+    else
+    {
+        SendResponse(aDnsHeader, ErrorToResponseCode(aError), aMessageInfo);
+    }
+}
+
+void SrpServer::Start()
+{
+    otError error;
+
+    VerifyOrExit(!IsRunning(), error = OT_ERROR_ALREADY);
+
+    SuccessOrExit(error = mSocket.Open(HandleUdpReceive, this));
+    SuccessOrExit(error = mSocket.Bind(0));
+
+    SuccessOrExit(error = PublishService());
+
+    otLogInfoSrp("SRP server: starts listening on port %hu", mSocket.GetSockName().mPort);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogCritSrp("failed to start SRP server: %s", otThreadErrorToString(error));
+        // Cleanup any resources we may have allocated.
+        Stop();
+    }
+}
+
+void SrpServer::Stop()
+{
+    VerifyOrExit(IsRunning());
+
+    UnpublishService();
+
+    while (!mHosts.IsEmpty())
+    {
+        Host::Destroy(mHosts.Pop());
+    }
+
+    while (!mOutstandingUpdates.IsEmpty())
+    {
+        UpdateMetadata::Destroy(mOutstandingUpdates.Pop());
+    }
+
+    mLeaseTimer.Stop();
+    mOutstandingUpdatesTimer.Stop();
+
+    otLogInfoSrp("SRP server: stops listening on %hu", mSocket.GetSockName().mPort);
+    IgnoreError(mSocket.Close());
+
+exit:
+    return;
+}
+
+void SrpServer::HandleNotifierEvents(Events aEvents)
+{
+    VerifyOrExit(mEnabled);
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
+    }
+
+exit:
+    return;
+}
+
+otError SrpServer::PublishService()
+{
+    otError       error;
+    const uint8_t serviceData[] = {kThreadServiceTypeSrpServer};
+    uint8_t       serverData[sizeof(uint16_t)];
+
+    OT_ASSERT(mSocket.IsBound());
+
+    *reinterpret_cast<uint16_t *>(serverData) = Encoding::BigEndian::HostSwap16(mSocket.GetSockName().mPort);
```

Comment:
Use `Encoding:BigEndian::WriteUint16()` instead. 

In general, casting a pointer to a buffer as `uint16_t *` and then reading/writing to it is not safe (will not behave correctly on all toolchains). Note that the `serverData` being a byte array has no alignment requirement and can start from any memory address, but `uint16_t` may have alignment requirement.